### ISSUE TITLE
Restore service monitoring in all pypi based images

### ISSUE THREAD

**Issue by vkuznet** [dmwm/WMCore#11454]
**Impact of the bug**
I discovered that when we moved to use pypi images we did not propagate `process_export` process within a pod. This leads to lost of service's specific metrics which Prometheus should scrape for service dashboards.

**Describe the bug**
In RPM based images we had `monitor.sh` script in docker images, e.g.
- [reqmgr2ms/monitor.sh](https://github.com/dmwm/CMSKubernetes/blob/master/docker/reqmgr2ms/monitor.sh)
- [reqmgr2ms-unmerged/monitor.sh](https://github.com/dmwm/CMSKubernetes/blob/master/docker/reqmgr2ms-unmerged/monitor.sh)
- etc.
Each `monitor.sh` script provides `process_exporter` process on service specific port which are used by Prometheus to scrape service metrics, e.g. see [services/reqmgr2ms-unmerged.yaml](https://github.com/dmwm/CMSKubernetes/blob/master/kubernetes/cmsweb/services/reqmgr2ms-unmerged-t1.yaml#L78) for how we set Promethues scrapes (other services have similar line and dedicated port).

**How to reproduce it**
All pypi docker files does not have `monitor.sh` script and/or `process_exporter` executable.

**Expected behavior**
We should put back either `monitor.sh` script or `process_exporter` into pypi based images and use them at `run.sh` and `manage`. Therefore, the fix will require the following:
- put back `monitor.sh` script and adjust it accordingly for each service Dockerfile
- `COPY` `process_exporter` from `exporters` image to `dmwm-base` image such that each service image will have it
- adjust `run.sh` and `manage` scripts to start `monitor.sh` one within a pod

**Additional context and error message**

**Comment by vkuznet** [comment]
The fix PR can be found here: https://github.com/dmwm/CMSKubernetes/pull/1301

I built and uploaded new `dmwm-base` image with these changes. The new tag is `registry.cern.ch/cmsweb/dmwm-base:pypi-20230125`.

**Comment by vkuznet** [comment]
The PR for using new `dmwm-base` image tag can be found here: https://github.com/dmwm/CMSKubernetes/pull/1302

**Comment by vkuznet** [comment]
I suggest to keep this issue open until we'll deploy and verify monitoring part of pypi based images in testbed or production.

**Comment by amaltaro** [comment]
@vkuznet Valentin, all the WMCore central services have just been upgraded to `2.1.8rc2`, which uses the latest dmwm-base docker image.

The new image brings in the monitor.sh script, but I failed to find its process in a couple of pods that I connected. Could you please double-check this?

**Comment by vkuznet** [comment]
I checked and we still have a problem due to order in logic of run.sh script, see https://github.com/dmwm/CMSKubernetes/blob/master/docker/pypi/dmwm-base/run.sh#L77-L83, i.e. we start monitor.sh script before we start the wmc-httpd daemon. And, according to monitor.sh logic, https://github.com/dmwm/CMSKubernetes/blob/master/docker/pypi/dmwm-base/monitor.sh#L7, we first check presence of wmc-httd daemon. Therefore, to make monitor.sh to start the appropriate exporter we should change the order in run.sh script to first start wmc-http daemon and then start the monitor script. The current order came from original RPM based design which start monitor script after we call the manage script. I made necessary changes over here: https://github.com/dmwm/CMSKubernetes/pull/1328 Please review and we'll need to rebuild base image one more time to fix the problem.

**Comment by vkuznet** [comment]
We have good and bad news :)
- the good news, the process exporter has started in dmwm pods and monitor.sh script runs fine, but
- the bad news, since many MS services uses `-` in their app name, the process exporter fails with the following error:
```
panic: descriptor Desc{fqName: "process_exporter_ms-monitor_process_cpu_seconds_total", help: "Total user and system CPU time spent in seconds (process collector)", constLabels: {}, variableLabels: []} is invalid: "process_exporter_ms-monitor_process_cpu_seconds_total" is not a valid metric name
....
```
which means that dashes are not allowed in metric names (since dash is interpreted as minus sign). I fixed issue with this PR https://github.com/dmwm/CMSKubernetes/pull/1331 and it means we'll need once again after its merge rebuild base image, and then use another PR to rebuild all MS images with new base image tag.

**Comment by vkuznet** [comment]
Finally, I"m happy to report that after deploying `2.2.0rc3` release version images to testbed we got back MS service monitoring plots. All of them can be found in this dashboard:
https://monit-grafana.cern.ch/d/JIz-EqvVz/wmcore-ms-services?orgId=11&refresh=1m&var-env=k8s-preprod&var-service=All&var-app=All&from=now-15m&to=now
where you'll need to select proper env (e.g. k8s-preprod) and your service (you may select multiple or all). I can see now metrics for testbed and when we put this into production they will appear as well.

@amaltaro , please have a look and we may close this ticket.

**Comment by amaltaro** [comment]
Yes, it looks like this issue has finally been resolved. I noticed that the "service" drop down menu starts with "process_exporter_*" though. One can clearly still see to which service each of those relate to, but if we had an alias/renaming to map to the actual service, that would be better. I think we can close this ticket though, doing so.

**Comment by vkuznet** [comment]
Yes, it should be fixed at CMS monitoring side, I created appropriate ticket [CMSMONIT-530](https://its.cern.ch/jira/browse/CMSMONIT-530)

**Comment by amaltaro** [comment]
As discussed in this ticket:
https://its.cern.ch/jira/browse/CMSMONIT-540

the reason for some of the alerts for ReqMgr2 is apparently associated to the fact that not all of the ReqMgr2 PODs are actually running `process_exporter`. Out of all the 6 reqmgr2-* PODs, I found one that didn't have it running (not clear to me why not!)

On what concerns reqmon, I confirm that process_exporter is not running and there is an error message when looking into the kubernetes logs. I will provide a fix to the dmwm-base image later today/tomorrow.

**Comment by amaltaro** [comment]
I have just created some test containers with tag `pypi-20230525-alan` and tested it with reqmgr2/reqmon/global-workqueue and they are now all starting the process_exporter and the prefix is correctly using the app name.

**Comment by amaltaro** [comment]
Hopefully this was the last fix on this topic, provided with:
https://github.com/dmwm/CMSKubernetes/pull/1381

and the PR mentioned in it.
